src/api/MainSolver.h

```diff
@@ -68,136 +51,162 @@ const sstat s_Error = toSstat( 2);
+        FrameId id;
+
     public:
-        PFRef last() const { assert(frames.size() > 0); return frames.last(); }
+        FrameId getId() const { return id; }
+        int size() const { return formulas.size(); }
+        void push(PTRef tr) { formulas.push(tr); }
+        PTRef operator[](int i) const { return formulas[i]; }
+        vec<PTRef> formulas;
+        bool unsat; // If true then the stack of frames with this frame at top is UNSAT
+
+        PushFrame(PushFrame const &) = delete;
+        PushFrame(PushFrame &&) = default;
+        explicit PushFrame(uint32_t id) : id({id}), unsat(false) {}
+    };
 
-        std::size_t size() const { return frames.size_(); }
+    class AssertionStack {
+    private:
+        std::vector<PushFrame> frames;
+        uint32_t frameId = 0;
 
-        const vec<PFRef>& getFrameReferences() const { return frames; }
+    public:
+        [[nodiscard]] PushFrame const & last() const {
+            assert(not frames.empty());
+            return frames.back();
+        }
+        [[nodiscard]] PushFrame & last() {
+            assert(not frames.empty());
+            return frames.back();
+        }
 
-        PFRef getFrameReference(std::size_t i) const { return frames[i]; }
+        [[nodiscard]] std::size_t frameCount() const { return frames.size(); }
 
-        std::size_t getSimplifiedUntil() const { return simplified_until; }
+        PushFrame const & operator[](std::size_t i) const {
+            assert(i < frames.size());
+            return frames[i];
+        }
+        PushFrame & operator[](std::size_t i) {
+            assert(i < frames.size());
+            return frames[i];
+        }
 
-        void setSimplifiedUntil(std::size_t n) { simplified_until = n; }
+        void push() { frames.emplace_back(frameId++); }
 
-        void push(PFRef frame_ref) { frames.push(frame_ref); }
+        void pop() { frames.pop_back(); }
 
-        void pop() {
-            frames.pop();
-            if (simplified_until > size()) { simplified_until = size(); }
+        void add(PTRef fla) {
+            assert(frameCount() > 0);
+            last().push(fla);
+        }
+    };
+
+    class Substitutions {
+        std::vector<Logic::SubstMap> perFrameSubst;
+
+    public:
+        void push() { perFrameSubst.emplace_back(); }
+        void pop() { perFrameSubst.pop_back(); }
+
+        void set(std::size_t level, Logic::SubstMap && subs) {
+            perFrameSubst.at(level) = std::move(subs);
+        }
+
+        Logic::SubstMap current() {
+            Logic::SubstMap allSubst;
+            for (auto const & subs : perFrameSubst) {
+                for (PTRef key : subs.getKeys()) {
+                    assert(not allSubst.has(key));
+                    allSubst.insert(key, subs[key]);
+                }
+            }
+            return allSubst;
+        }
+    };
+/** Actual MainSolver members **/
+protected:
```

Comment:
Fix tabulation(and below/above)

Reply:
I believe indentation is correct, no?
This (and below) belongs to the `MainSolver` class.
Above you have some nested classes, so their members are indented more.

Reply:
Yup, sry, my bad (was a bit confused with deetions and additions), but there is duplicate `public` lines 182, 205